wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift

```diff
@@ -24,33 +24,58 @@ import CoreCryptoSwift
 public protocol SafeCoreCryptoProtocol {
     func perform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T
     func unsafePerform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T
+    func mlsInit(clientID: String) throws
 }
 
-public struct SafeCoreCrypto: SafeCoreCryptoProtocol {
+public class SafeCoreCrypto: SafeCoreCryptoProtocol {
     public enum CoreCryptoSetupFailure: Error, Equatable {
         case failedToGetClientIDBytes
     }
 
     private let coreCrypto: CoreCryptoProtocol
     private let safeContext: SafeFileContext
+    private var didInitializeMLS = false
 
-    public init(coreCryptoConfiguration config: CoreCryptoConfiguration) throws {
-        guard let clientID = config.clientIDBytes() else {
+    public convenience init(coreCryptoConfiguration config: CoreCryptoConfiguration) throws {
+        guard let clientID = config.clientIDBytes else {
             throw CoreCryptoSetupFailure.failedToGetClientIDBytes
         }
+
         let coreCrypto = try CoreCrypto(
             path: config.path,
             key: config.key,
             clientId: clientID,
             entropySeed: nil
         )
 
-        self.init(coreCrypto: coreCrypto, coreCryptoConfiguration: config)
+        self.init(coreCrypto: coreCrypto, path: config.path)
+        didInitializeMLS = true
+    }
+
+    public convenience init(path: String, key: String) throws {
+        let coreCrypto = try CoreCrypto.deferredInit(
+            path: path,
+            key: key,
+            entropySeed: nil
+        )
+
+        self.init(coreCrypto: coreCrypto, path: path)
+    }
+
+    public func mlsInit(clientID: String) throws {
+        guard !didInitializeMLS else { return }
+
+        guard let clientIdBytes = ClientId(from: clientID) else {
+            throw CoreCryptoSetupFailure.failedToGetClientIDBytes
+        }
+
+        try coreCrypto.mlsInit(clientId: clientIdBytes)
+        didInitializeMLS = true
     }
 
-    init(coreCrypto: CoreCryptoProtocol, coreCryptoConfiguration: CoreCryptoConfiguration) {
+    init(coreCrypto: CoreCryptoProtocol, path: String) {
```

Comment:
maybe we need to be more  specific about the path here since we modify the path afterwards

Reply:
I'm not sure what else would make sense, do you have a suggestion?

Reply:
`databasePath`?
or coredataStackDirectory and move the `let directoryPathUrl = URL(fileURLWithPath: path).deletingLastPathComponent()` outside of init

Reply:
changed to `databasePath`